[FSDP] Use _init_from_local_tensor to create ShardedTensor to avoid communication overhead

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #77628
* __->__ #82911

FSDP originally uses `_init_from_local_shards_and_global_metadata()` to create a ShardedTensor for sharded_state_dict(). We have seen some non-trivial overhead if the number of tensors is large. Using `_init_from_local_shards_and_global_metadata ` can significantly reduce the overhead. For a model with ~250 tensors in the state_dict trained with 16 GPUs, the original `sharded_state_dict` takes ~1.7 seconds and this PR reduces the overhead to ~0.6 seconds.

Differential Revision: [D38452170](https://our.internmc.facebook.com/intern/diff/D38452170/)

fegin committed:
[FSDP] Use _init_from_local_tensor to create ShardedTensor to avoid communication overhead

FSDP originally uses `init_from_local_shards()` to create a ShardedTensor for sharded_state_dict(). We have seen some non-trivial overhead if the number of tensors is large. Using `_init_from_from_local_tensor` can significantly reduce the overhead. For a model with ~250 tensors in the state_dict trained with 16 GPUs, the original `sharded_state_dict` takes ~1.7 seconds and this PR reduces the overhead to ~0.6 seconds.

Differential Revision: [D38452170](https://our.internmc.facebook.com/intern/diff/D38452170/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Use _init_from_local_tensor to create ShardedTensor to avoid communication overhead"


FSDP originally uses `init_from_local_shards()` to create a ShardedTensor for sharded_state_dict(). We have seen some non-trivial overhead if the number of tensors is large. Using `_init_from_from_local_tensor` can significantly reduce the overhead. For a model with ~250 tensors in the state_dict trained with 16 GPUs, the original `sharded_state_dict` takes ~1.7 seconds and this PR reduces the overhead to ~0.6 seconds.

Differential Revision: [D38452170](https://our.internmc.facebook.com/intern/diff/D38452170/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Use _init_from_local_tensor to create ShardedTensor to avoid communication overhead"


FSDP originally uses `init_from_local_shards()` to create a ShardedTensor for sharded_state_dict(). We have seen some non-trivial overhead if the number of tensors is large. Using `_init_from_from_local_tensor` can significantly reduce the overhead. For a model with ~250 tensors in the state_dict trained with 16 GPUs, the original `sharded_state_dict` takes ~1.7 seconds and this PR reduces the overhead to ~0.6 seconds.

Differential Revision: [D38452170](https://our.internmc.facebook.com/intern/diff/D38452170/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Use _init_from_local_tensor to create ShardedTensor to avoid communication overhead"


FSDP originally uses `init_from_local_shards()` to create a ShardedTensor for sharded_state_dict(). We have seen some non-trivial overhead if the number of tensors is large. Using `_init_from_from_local_tensor` can significantly reduce the overhead. For a model with ~250 tensors in the state_dict trained with 16 GPUs, the original `sharded_state_dict` takes ~1.7 seconds and this PR reduces the overhead to ~0.6 seconds.

Differential Revision: [D38452170](https://our.internmc.facebook.com/intern/diff/D38452170/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Use _init_from_local_tensor to create ShardedTensor to avoid communication overhead"


FSDP originally uses `init_from_local_shards()` to create a ShardedTensor for sharded_state_dict(). We have seen some non-trivial overhead if the number of tensors is large. Using `_init_from_from_local_tensor` can significantly reduce the overhead. For a model with ~250 tensors in the state_dict trained with 16 GPUs, the original `sharded_state_dict` takes ~1.7 seconds and this PR reduces the overhead to ~0.6 seconds.

Differential Revision: [D38452170](https://our.internmc.facebook.com/intern/diff/D38452170/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Use _init_from_local_tensor to create ShardedTensor to avoid communication overhead"


FSDP originally uses `init_from_local_shards()` to create a ShardedTensor for sharded_state_dict(). We have seen some non-trivial overhead if the number of tensors is large. Using `_init_from_from_local_tensor` can significantly reduce the overhead. For a model with ~250 tensors in the state_dict trained with 16 GPUs, the original `sharded_state_dict` takes ~1.7 seconds and this PR reduces the overhead to ~0.6 seconds.

Differential Revision: [D38452170](https://our.internmc.facebook.com/intern/diff/D38452170/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Use _init_from_local_tensor to create ShardedTensor to avoid communication overhead"


FSDP originally uses `init_from_local_shards()` to create a ShardedTensor for sharded_state_dict(). We have seen some non-trivial overhead if the number of tensors is large. Using `_init_from_from_local_tensor` can significantly reduce the overhead. For a model with ~250 tensors in the state_dict trained with 16 GPUs, the original `sharded_state_dict` takes ~1.7 seconds and this PR reduces the overhead to ~0.6 seconds.

Differential Revision: [D38452170](https://our.internmc.facebook.com/intern/diff/D38452170/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/shard_utils.py:
```diff
@@ -188,3 +194,56 @@ def _gather_state_dict(
+    else:
+        local_shards = []
+
+    # Create a ShardedTensor without invoking communnication.
```
nit:
```suggestion
    # Create a ShardedTensor without invoking communication.
```

awgu commented on torch/distributed/fsdp/shard_utils.py:
```diff
@@ -188,3 +194,56 @@ def _gather_state_dict(
+            pin_memory=tensor.is_pinned(),
+        )
+    )
+    return ShardedTensor._init_from_local_shards_and_global_metadata(
```
It looks like this does not use `_init_from_local_tensor()` (anymore)? Maybe we should update the PR title/summary.

awgu commented on torch/distributed/fsdp/shard_utils.py:
```diff
@@ -188,3 +194,56 @@ def _gather_state_dict(
+    Shard a tensor to chunks along the first dimension. The local rank will gets its
+    corresponding chunk as the local shard to create a ShardedTensor.
+    """
+    chunks = tensor.chunk(world_size)
```
nit: Do you think it is worth explicitly passing `dim=0` into `chunk()`? It might make it more clear that `chunk()` is chunking along the 0th dimension.

awgu commented on torch/distributed/fsdp/shard_utils.py:
```diff
@@ -188,3 +194,56 @@ def _gather_state_dict(
+        local_shard = chunks[rank].clone()
+        offsets = [0 for _ in tensor.size()]
+        offsets[0] = math.ceil(tensor.size()[0] / world_size) * rank
+        local_shards = [Shard.from_tensor_and_offsets(local_shard, offsets, rank)]
```
Conceptual check: `local_shards` here has length 1? (So overall, it is either length 1 if this rank has a chunk, or length 0 if this rank does not have a chunk?)

fegin replied:
That's correct!

awgu commented on test/distributed/fsdp/test_shard_utils.py:
```diff
@@ -185,3 +186,20 @@ def get_offsets(tensor, shard):
+
+    @skip_if_lt_x_gpu(2)
+    def test_create_chunk_sharded_tensor(self):
+        for size in ((1,), (1, 6), (12,), (12, 6), (25,), (25, 6)):
```
Do you think testing at most 2 dimensions suffices? I guess it should be fine since you are only sharding along dim 0 so as long as there is a second dim, it should be okay.

fegin replied:
Unless we change the implementation to chunk in a different dimension, 2 dimensions test should be enough.

awgu reviewed (approved):
LGTM. I left some nits, but I am stamping to unblock.

fegin commented:
@pytorchbot merge

fegin committed:
Update on "[FSDP] Use _init_from_local_tensor to create ShardedTensor to avoid communication overhead"


FSDP originally uses `_init_from_local_shards_and_global_metadata()` to create a ShardedTensor for sharded_state_dict(). We have seen some non-trivial overhead if the number of tensors is large. Using `_init_from_local_shards_and_global_metadata ` can significantly reduce the overhead. For a model with ~250 tensors in the state_dict trained with 16 GPUs, the original `sharded_state_dict` takes ~1.7 seconds and this PR reduces the overhead to ~0.6 seconds.

Differential Revision: [D38452170](https://our.internmc.facebook.com/intern/diff/D38452170/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Use _init_from_local_tensor to create ShardedTensor to avoid communication overhead"


FSDP originally uses `_init_from_local_shards_and_global_metadata()` to create a ShardedTensor for sharded_state_dict(). We have seen some non-trivial overhead if the number of tensors is large. Using `_init_from_local_shards_and_global_metadata ` can significantly reduce the overhead. For a model with ~250 tensors in the state_dict trained with 16 GPUs, the original `sharded_state_dict` takes ~1.7 seconds and this PR reduces the overhead to ~0.6 seconds.

Differential Revision: [D38452170](https://our.internmc.facebook.com/intern/diff/D38452170/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Use _init_from_local_tensor to create ShardedTensor to avoid communication overhead"


FSDP originally uses `_init_from_local_shards_and_global_metadata()` to create a ShardedTensor for sharded_state_dict(). We have seen some non-trivial overhead if the number of tensors is large. Using `_init_from_local_shards_and_global_metadata ` can significantly reduce the overhead. For a model with ~250 tensors in the state_dict trained with 16 GPUs, the original `sharded_state_dict` takes ~1.7 seconds and this PR reduces the overhead to ~0.6 seconds.

Differential Revision: [D38452170](https://our.internmc.facebook.com/intern/diff/D38452170/)

[ghstack-poisoned]
